### ISSUE TITLE
Backward of maxmin-cnn layer

### ISSUE THREAD

**Issue by hiyijian** [karandesai-96/maxmin-cnn#1]
hi @karandesai-96,
  Thank you for sharing the nice and clean idea with us. I have readed the Readme.md, code and blog post carefully, But not yet your paper. I think the backward of maxmin-cnn layer will be diffrient from regular one, since it's output has double amout of channels. Am I right?  However, I didnt find any backward code  in your script of keras_maxmin_impl.py or description from above materials. Would you please to explain it more?

**Comment by karandesai-96** [comment]
Hi @hiyijian thanks for your interest in my repository and blog. I should mention that it is not my paper, it is a very recent paper which doesn't have much related references yet on internet. 

> I think the backward of maxmin-cnn layer will be different from regular one, since it's output has double amount of channels.

The output of maxmin-cnn layer has double channels but the **extra** channels added do not have any trainable weights associated with them. So assume the computation graph:

* The gradient flows backwards from the ReLU activation through all (say `2 * n`) channels. during backward pass.
* For the first half ( `n` ) channels, there are trainable weights associated with the filters and the gradient flows backwards same as a normal conv layer.
* For the second half ( `n` ) channels, there are **no trainable weights** but only constant co-efficients with value -1 coming out of first half of channels. So the gradient flow stops here itself.

The overall meaning being, this negation and concatenation logic happens while forward pass / prediction, but it doesn't alter the training.

**NOTE**: Hypothetically (because this is not what happens in behind the curtains of Keras)  you may look at maxmin layer as a simple conv layer with double filters than original but assuming a constraint that two channels have a shared filter weights - which are exactly negative of each other. Weight updates in both filters take place by same amount, but different signs.

**Comment by karandesai-96** [comment]
> However, I didnt find any backward code in your script of keras_maxmin_impl.py or description from above materials.

Keras depends on either **Theano or Tensorflow**, both of which represents our forward and backward passes of Neural Networks as nodes and branches of a computational graph. So we only need to define the logic of each layer from input through output. The flow of gradients in either directions is handled by Theano or Tensorflow itself.

I did not require to write the backward code due to this reason, the Keras Documentation for writing a custom layer class say that we need to override the "call" method to put the layer logic inside.

**Reference:** https://keras.io/layers/writing-your-own-keras-layers/

**Comment by karandesai-96** [comment]
@hiyijian: I hope my explanation is clear to you, I'll be happy to clarify it further if there's any confusion about the explanation. Thanks for raising this issue, the paper hasn't explained this in terms of nodes of computational graph, it didn't require to as it is specific to these Python libraries, but this will be useful for other having similar doubts :smile:

**Comment by hiyijian** [comment]
Much thanks. It is clear now.

**Comment by hiyijian** [comment]
PS: the model size will be quite larger than the one using regular conv layer, since params of conv layer will increase by 2^n？

**Comment by karandesai-96** [comment]
> PS: the model size will be quite larger than the one using regular conv layer, since params of conv layer will increase 2^n in a layer by layer fashion？

Even I thought so, but later I realized that due to absence of trainable weights for half of channels in a maxmin conv layer, the parameters do not increase that much. **_The output channels get doubled so they make the parameters for next layer double (due to increased filter channel depth). But that doesn't grow later._**

I verified it by hand calculations, and Keras showed me the same.. Please check out my notebook for **CIFAR 10** dataset. I have printed `summary` of both models near the end. Classinal CNN has approximately 116k parameters while MaxMin CNN has 230k parameters, while are approximately only double even on introduction of three MaxMin cov layers.

**Comment by karandesai-96** [comment]
In case if multiple notebooks get added in this repo, for the sake of completeness here is the screenshot: 

![image](https://cloud.githubusercontent.com/assets/10494087/21565734/6109dc68-cec1-11e6-87e7-833b66cfcdbc.png)

This fact was very deceptive for me as well but later I realized that the exponential growth is not happening actually..

**Comment by hiyijian** [comment]
you are right. thanks for your time
